### PR TITLE
Hide the “Add to Cart” button if a bid has been placed on an item by …

### PR DESCRIPTION
**Pull request by vikas-cis**
#984

### CODE REVIEW

**Comment by kewde** [comment]
You are effectively caching the hashes, please use the same structure found in the services of market-cache.

**Comment by vikas-cis** [comment]
@kewde please check now and let me know if that looks good.

Thanks

File: src/app/core/market/market-cache/add-to-cart-cache.service.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+  private log: any = Log.create('add-to-cart-cache.service id:' + Math.floor((Math.random() * 1000) + 1));
+  public orders: Array<any> = new Array();;
+  public hashes: Array<string>
+  constructor(
+    private marketState: MarketStateService
+  ) {
+    this.update();
+    // subscribe to changes
+    this.getBids().subscribe(orders => {
```
Comment (kewde) [inline src/app/core/market/market-cache/add-to-cart-cache.service.ts]:
```
takeWhile(() => !this.destroyed)
```

File: src/app/market/listings/listing-item/listing-item.component.ts [hunk 2]
```diff
@@ -15,16 +14,21 @@ import { PreviewListingComponent } from '../preview-listing/preview-listing.comp
   templateUrl: './listing-item.component.html',
   styleUrls: ['./listing-item.component.scss']
 })
-export class ListingItemComponent {
+export class ListingItemComponent implements OnInit {
   @Input() listing: Listing;
-
+  // Input hashes
+  @Input() hashes: Array<any> = new Array();
```
Comment (kewde) [inline src/app/market/listings/listing-item/listing-item.component.ts]: rm

File: src/app/market/listings/listing-item/listing-item.component.ts [hunk 3]
```diff
@@ -15,16 +14,21 @@ import { PreviewListingComponent } from '../preview-listing/preview-listing.comp
...
   constructor(private dialog: MatDialog,
-              private cartService: CartService,
               private favoritesService: FavoritesService,
               private snackbar: SnackbarService,
               private marketState: MarketStateService) {
   }
 
+  ngOnInit() {
```
Comment (kewde) [inline src/app/market/listings/listing-item/listing-item.component.ts]: rm

File: src/app/market/buy/buy.component.ts [hunk 4]
```diff
@@ -49,6 +49,7 @@ export class BuyComponent implements OnInit {
       const temp: Listing[] = [];
       favorites.forEach(favorite => {
         this.listingService.get(favorite.listingItemId).take(1).subscribe(listing => {
+          console.log(listing);
```
Comment (kewde) [inline src/app/market/buy/buy.component.ts]: rm

File: src/app/core/market/api/template/template.model.ts [hunk 5]
```diff
@@ -16,7 +16,7 @@ export class Template {
   public internationalTotal: Amount = new Amount(0);
   public totalAmount: Amount = new Amount(0);
   public memo: string = '';
-
+  public hashes: Array<any> = new Array();
```
Comment (kewde) [inline src/app/core/market/api/template/template.model.ts]: rm

File: src/app/core/market/api/bid/bid.model.ts [hunk 6]
```diff
@@ -10,6 +10,8 @@ export class Bid {
   public listing: Listing;
   public listingItemId: number;
   public status: string;
+  public hashes: Array<any> = new Array();
```
Comment (kewde) [inline src/app/core/market/api/bid/bid.model.ts]: rm

File: src/app/core/market/market-cache/add-to-cart-cache.service.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+import { Bid } from '../api/bid/bid.model';
+import { MarketStateService } from 'app/core/market/market-state/market-state.service';
+import { Listing } from 'app/core/market/api/listing/listing.model';
+
+@Injectable()
+export class AddToCartCacheService {
+
+  private log: any = Log.create('add-to-cart-cache.service id:' + Math.floor((Math.random() * 1000) + 1));
+  public orders: Array<any> = new Array();;
+  public hashes: Array<string>
```
Comment (kewde) [inline src/app/core/market/market-cache/add-to-cart-cache.service.ts]: rm

**Comment by kewde** [comment]
Don't forget to fix the test.

**Comment by kewde** [comment]
@vikas-cis fix test

File: src/app/core/market/market-cache/add-to-cart-cache.service.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
+import { Injectable, OnDestroy } from '@angular/core';
+import * as _ from 'lodash';
```
Comment (kewde) [inline src/app/core/market/market-cache/add-to-cart-cache.service.ts]: rm lodash if not used

File: src/app/market/listings/listings.component.html [hunk 9]
```diff
@@ -108,7 +108,7 @@
     <ng-container *ngFor="let page of pages">
       <ng-container *ngFor="let item of page.listings">
         <mat-card class="listing" fxFlex="0 0 calc(100% / 4 - 30px)">
-          <app-listing-item [listing]="item"></app-listing-item>
+          <app-listing-item [listing]="item" [hashes]="hashes"></app-listing-item>
```
Comment (kewde) [inline src/app/market/listings/listings.component.html]: rm hashes?

File: src/app/market/listings/preview-listing/preview-listing.component.html [hunk 10]
```diff
@@ -75,12 +75,8 @@
...
     </div>
   </div>
-  <div class="right" *ngIf="!data?.listing?.object.ListingItemTemplate?.hash">
-    <app-favorite [listing]="data?.listing"></app-favorite>
-
-    <button mat-raised-button color="primary" (click)="addToCart(data?.listing)">
-      <mat-icon fontSet="partIcon" fontIcon="part-check"></mat-icon>
-      Add to cart
-    </button>
+  <div class="right" fxLayout fxLayoutAlign="end center" *ngIf="!data?.listing?.object.ListingItemTemplate?.hash">
```
Comment (kewde) [inline src/app/market/listings/preview-listing/preview-listing.component.html]: Doesn't there exist a 'hash' object in the listing model?

try rewriting it as such:
*ngIf="!data?.listing?.hash"

File: src/app/market/shared/addtocart/add-to-cart.component.ts [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+})
+export class AddToCartComponent implements OnInit {
+
+  private log: any = Log.create('add-to-cart.component id:' + Math.floor((Math.random() * 1000) + 1));
+
+  @Input() listing: Listing;
+
+  constructor(
+    private cartService: CartService,
+    private cache: AddToCartCacheService
```
Comment (kewde) [inline src/app/market/shared/addtocart/add-to-cart.component.ts]: Store the AddToCartCacheService as public variable named 'cache' in the CartService.
Then refer to it as:
this.cartService.cache.isBidded(..)

**Comment by kewde** [comment]
Lint errors

File: src/app/market/listings/listings.component.html (hunk 9 above)
Comment (vikas-cis) [reply src/app/market/listings/listings.component.html]: yes correct (y)

File: src/app/market/listings/preview-listing/preview-listing.component.html (hunk 10 above)
Comment (vikas-cis) [reply src/app/market/listings/preview-listing/preview-listing.component.html]: I thought we are checking the template related hash and guess both are same so we can make that the above :)

File: src/app/market/shared/addtocart/add-to-cart.component.ts (hunk 11 above)
Comment (vikas-cis) [reply src/app/market/shared/addtocart/add-to-cart.component.ts]: yup 

thanks

File: src/app/market/sell/add-item/add-item.component.html [hunk 12]
```diff
@@ -8,8 +8,8 @@
         <mat-icon class="icon" fontSet="partIcon" fontIcon="part-select"></mat-icon>
 
         <!--hide input -->
-        <input style="display: none;" type="file"
-               accept="image/*" multiple="true" id="fileInput">
+        <!-- <input style="display: none;" type="file"
```
Comment (kewde) [inline src/app/market/sell/add-item/add-item.component.html]: NO!

You are again breaking this, what the heck...

Comment (kewde) [reply src/app/market/sell/add-item/add-item.component.html]: It breaks the test by the way.

Other than that - this PR is mergeable.

Comment (vikas-cis) [reply src/app/market/sell/add-item/add-item.component.html]: @kewde if it is display:none then how it breaks the component tho ?

Thanks

Comment (vikas-cis) [reply src/app/market/sell/add-item/add-item.component.html]: tho I have changed that to class rather commenting it out 

Thanks

**Comment by vikas-cis** [comment]
@kewde its Mergeable now 

Thanks

**Pull request merged**